### PR TITLE
snap: fix interface bindings on implicit hooks

### PR DESCRIPTION
**Pull request by zyga**
An implicit hook is one that exists in the filesystem but is not
mentioned in the meta/snap.yaml file.

An interface binding is an association of a plug or slot definition with
an app or hook.

Plugs and slots can be defined either at the top level using the full
syntax (comprised of key-value pairs) or at an app or hook level using
the abbreviated syntax (with the interface name alone).

When an application or hook references a top-level plug or slot, or
defines a new plug or slot named after the interface name, that act
automatically scopes the interface to a particular application or hook.

This feature allows snap developers to implement a form of privilege
separation, where particular applications or hooks have a subset of the
permissions resulting from the set of interface connections.

In theory implicit hook definitions are no different from explicit hook
definitions. Due to a bug however an implicitly defined hook would get
*all* the top-level plugs and slots defined in a given snap, even if
those interfaces were already scoped to some applications. This would
not happen if said hook was explicitly defined in the YAML and not
bound to any plug or slot.

This patch fixes this by not considering top-level plugs and slots that
already have app or hook associations. Such interfaces must be already
scoped to a particular app or hook so they would not be assigned to an
hook if it was explicitly defined.

Fixes: https://bugs.launchpad.net/snapd/+bug/1824557
Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: snap/info_snap_yaml.go [hunk 1]
```diff
@@ -503,9 +503,12 @@ func bindUnboundSlots(snap *Info) {
...
 func bindImplicitHooks(snap *Info) {
-	for _, plug := range snap.toplevelPlugs {
-		for hookName, hook := range snap.Hooks {
-			if hook.Explicit {
+	for hookName, hook := range snap.Hooks {
+		if hook.Explicit {
+			continue
+		}
+		for _, plug := range snap.toplevelPlugs {
+			if len(plug.Apps) > 0 || len(plug.Hooks) > 0 {
```
Comment (pedronis) [inline snap/info_snap_yaml.go]: the loop changes plug.Hooks itself that is used in the condition here so something doesn't look right, this probably needs more tests

Comment (zyga) [reply snap/info_snap_yaml.go]: Ah, indeed, good point!

Comment (zyga) [reply snap/info_snap_yaml.go]: This is addressed now.

File: snap/info_snap_yaml.go [hunk 2]
```diff
@@ -503,9 +503,23 @@ func bindUnboundSlots(snap *Info) {
...
+	bound := make(map[string]bool, len(snap.Plugs)+len(snap.Slots))
 	for _, plug := range snap.toplevelPlugs {
-		for hookName, hook := range snap.Hooks {
-			if hook.Explicit {
+		if len(plug.Apps) > 0 || len(plug.Hooks) > 0 {
+			bound[plug.Name] = true
+		}
+	}
+	for _, slot := range snap.toplevelSlots {
+		if len(slot.Apps) > 0 || len(slot.Hooks) > 0 {
```
Comment (pedronis) [inline snap/info_snap_yaml.go]: There is still something not quite right I fear going on, it's a bit too late to check this here, which means again missing tests.  What we care is really the state of plugs and slots before we run bindUnboundPlugs/Slots.  Maybe the easiest is to replace toplevelPlugs/Slots with unboundPlugs/Slots and compute them after  setHooksFromSnapYaml.

Comment (zyga) [reply snap/info_snap_yaml.go]: I agree, I see the mistake now.

The mistake is this (symmetric for plus):
- a plug is defined at the top level
- it is not referenced anywhere so it binds to all apps and hooks
- it gets bound to an app or a hook and thus the referenced condition no longer holds true
- it does not get bound to an implicitly defined hook, **incorrectly so**

Comment (zyga) [reply snap/info_snap_yaml.go]: This is now fixed.

**Comment by zyga** [comment]
The commit message on 2393712b14089b7304906c4358209aace35157bc is enlightening.

**Comment by zyga** [comment]
Merged with permission from @mvo5

**Pull request merged**